### PR TITLE
Performance test as a service workflow

### PR DESCRIPTION
**Pull request by patrick-stephens**
Creates a packer image in GCP with various tools installed.
Template skeleton to run tests with compose and collect metrics.
Hooks for overriding various aspects.

The script is compose+monitoring and can be run by anyone like so as a one-liner without needing installation:
```
curl -L https://raw.githubusercontent.com/fluent/fluent-bit-ci/master/scripts/docker-compose-monitor.sh | bash
```
We provide configuration via environment variables:
* Operate on local compose stack or a git repo
* Run indefinitely or for a limited time
* Dump metrics after exit

The goal is to just provide some simple monitoring on top of a default compose stack.

For the two use cases we currently have, these require slightly different configs:
1. Threaded input testing - a custom fluent-bit.conf (`threaded on`) and a custom branch
   * Sample repo has a custom docker-compose.yml with the config loading it as we also may want to mess with env vars on the container too: https://github.com/patrick-stephens/perf-test-poc/tree/main/compose
2. Tracing support - a custom branch and an extra script to turn it on via the endpoint
   * Custom config uses a sidecar container to enable tracing: https://github.com/fluent/fluent-bit/tree/tap-trace-output/examples/tracing

Potentially we could build up a set of these example compose files as part of the normal work.

Examples
-------------

Running with threaded test (1), sample repo with sub-directory containing custom compose stack: https://github.com/patrick-stephens/perf-test-poc/tree/main/compose
```
export GIT_URL=https://github.com/patrick-stephens/perf-test-poc
export TEST_DIRECTORY=compose
export SERVICE_TO_MONITOR=fb-delta
unset GIT_REF
curl -L https://raw.githubusercontent.com/fluent/fluent-bit-ci/perf_test_workflow/scripts/docker-compose-monitor.sh | bash
```
fluentbit_input_records_total

Running with tracing test (2):
```
export GIT_URL=https://github.com/fluent/fluent-bit
export GIT_REF=tap-trace-output
export TEST_DIRECTORY=examples/tracing
export SERVICE_TO_MONITOR=fluent-bit
curl -L https://raw.githubusercontent.com/fluent/fluent-bit-ci/perf_test_workflow/scripts/docker-compose-monitor.sh | bash
```

Prometheus data loader
-------------------------------
Added a simple compose stack with Prometheus and Grafana that can load archived data from the monitoring script:
```
$ PROMETHEUS_DATA=/tmp/output/prom-data.tgz helpers/prometheus-snapshot-loader/run.sh
```
Screenshot from 2022-06-22 21-01-12

### CODE REVIEW

File: .github/workflows/run-test.yaml [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+        description: The baseline version of Fluent Bit to compare against.
+        required: false
+        type: string
+        default: fluent/fluent-bit:latest
+      ref:
+        description: The branch name to use for comparison against, this will be built as a container to compare with.
+        required: true
+        type: string
+        default: https://github.com/fluent/fluent-bit.git#master
+      custom-config-ref:
```
Comment (patrick-stephens) [inline .github/workflows/run-test.yaml:16]: My thinking here is we can either have a whole repo with a test case, or a repo with a set of test cases then the custom-environment can specify the correct context/compose stack.

File: .github/workflows/run-test.yaml [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            export OUTPUT_DIR=~/output; \
+            export FB_BASELINE_IMAGE=${{ github.event.inputs.baseline }}; \
+            export FB_DELTA_REF=${{ github.event.inputs.ref }}; \
+            /opt/fluent-bit-ci/templates/test/run.sh"
+        shell: bash
+
+      - name: Collect info
+        if: always()
+        run: |
+          gcloud compute scp --force-key-file-overwrite --recurse "$VM_NAME":~/output output
```
Comment (patrick-stephens) [inline .github/workflows/run-test.yaml:128]: Need to figure out the script command to add this as a comment to a PR later...

File: packer/templates/monitoring/prometheus/prometheus.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    scrape_interval: 5s
+    metrics_path: /api/v1/metrics/prometheus
+    static_configs:
+         - targets: ['fb-baseline:2020']
+
+  - job_name: 'fb-delta'
+    scrape_interval: 5s
+    metrics_path: /api/v1/metrics/prometheus
+    static_configs:
+         - targets: ['fb-delta:2020']
```
Comment (patrick-stephens) [inline packer/templates/monitoring/prometheus/prometheus.yml:28]: Configuring this with prometheus as an option is maybe something we want to do later...

If we provide a custom compose stack then we possibly can use the sd_dns_config approach to monitor for new compose services and just scrape them? TODO: investigate.

**Comment by patrick-stephens** [comment]
I'm still unhappy with the complexity involved. I'm going to try to simplify it further to a simple compose wrapper that gives you monitoring output.

**Comment by patrick-stephens** [comment]
Going to add a helper script as well (later) for Prometheus snapshot loading into a container too, likely with a Grafana stack as well to let people query it all.

File: .github/workflows/update-packer-vm.yaml [hunk 4]
```diff
@@ -0,0 +1,58 @@
+---
+name: Update packer images
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main
+      - master
```
Comment (niedbalski) [inline .github/workflows/update-packer-vm.yaml:8]: I think we only have main.

**Pull request merged**